integrated_channels/canvas/models.py

```diff
@@ -106,3 +121,97 @@ def get_content_metadata_transmitter(self):
         Return a ``CanvasContentMetadataTransmitter`` instance.
         """
         return CanvasContentMetadataTransmitter(self)
+
+
+@python_2_unicode_compatible
+class CanvasLearnerDataTransmissionAudit(models.Model):
+    """
+    The payload we sent to canvas at a given point in time for an enterprise course enrollment.
+
+    """
+    canvas_user_email = models.CharField(
+        max_length=255,
+        blank=False,
+        null=False
+    )
+
+    enterprise_course_enrollment_id = models.PositiveIntegerField(
+        blank=False,
+        null=False,
+        db_index=True
+    )
+
+    course_id = models.CharField(
+        max_length=255,
+        blank=False,
+        null=False,
+        help_text="The course run's key which is used to uniquely identify the course for Degreed."
+    )
+
+    course_completed = models.BooleanField(
+        default=True,
+        help_text="The learner's course completion status transmitted to Degreed."
+    )
+
+    completed_timestamp = models.CharField(
+        max_length=10,
+        help_text=(
+            'Represents the canvas representation of a timestamp: yyyy-mm-dd, '
+            'which is always 10 characters.'
+        )
+    )
+
+    # Request-related information.
+    status = models.CharField(max_length=100, blank=False, null=False)
+    error_message = models.TextField(blank=True)
+    created = models.DateTimeField(auto_now_add=True)
+    grade = models.CharField(max_length=255, null=True, blank=True)
+
+    class Meta:
+        app_label = 'canvas'
+
+    def __str__(self):
+        """
+        Return a human-readable string representation of the object.
+        """
+        return (
+            '<CanvasLearnerDataTransmissionAudit {transmission_id} for enterprise enrollment {enrollment}, '
+            'email {canvas_user_email}, and course {course_id}>'.format(
+                transmission_id=self.id,
+                enrollment=self.enterprise_course_enrollment_id,
+                canvas_user_email=self.canvas_user_email,
+                course_id=self.course_id
+            )
+        )
+
+    def __repr__(self):
+        """
+        Return uniquely identifying string representation.
+        """
+        return self.__str__()
+
+    def serialize(self, *args, **kwargs):  # pylint: disable=unused-argument
+        """
+        Return a JSON-serialized representation.
+
+        Sort the keys so the result is consistent and testable.
+
+        Can take the following keyword arguments:
+            - `enterprise_configuration`: used to get the `canvas_account_id` to be sent in the payload.
+
+        # TODO: When we refactor to use a serialization flow consistent with how course metadata
+        # is serialized, remove the serialization here and make the learner data exporter handle the work.
+        """
+        return json.dumps(self._payload_data(), sort_keys=True)
```

Comment:
oh I had to do this for Moodle too, even with exporter handling, so meeeehh.

Reply:
😒 so much debugging just to figure out I'm missing a specific required method in the model